File: Dynamic Programming/1937번 - 욕심쟁이 판다/1937.cpp
#include <iostream>

using namespace std;

int n;
int forest[501][501] = { 0, };
int dp[501][501] = { 0, };

int dx[4] = { 0, -1, 0, 1 };
int dy[4] = { -1, 0, 1, 0 };

void input() {
    cin >> n;
    for(int i = 0; i < n; i++) {
        for(int j = 0; j <n; j++) {
            cin >> forest[i][j];
        }
    }
}

int dfs(int x, int y) {
    if(dp[x][y]) return dp[x][y];
    dp[x][y] = 1;
    for(int i = 0; i < 4; i++) {
        int nx = x + dx[i];
        int ny = y + dy[i];
        if(nx >= 0 && nx < n && ny >= 0 && ny < n && forest[nx][ny] > forest[x][y]) {
            dp[x][y] = max(dp[x][y], dfs(nx, ny) + 1 );
        }
    }
    
    return dp[x][y];
}

void solve() {
    int result = 0;
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j <n; j++) {
            result = max(result, dfs(i, j));
        }
    }
    
    cout << result << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    input();
    solve();
    
    return 0;
}

File: Dynamic Programming/1937번 - 욕심쟁이 판다/1937.py
import sys

input = sys.stdin.readline
sys.setrecursionlimit(10**6)

n = int(input())
forest = [list(map(int, input().split())) for _ in range(n)]
dp = [[0] * n for _ in range(n)]

dx = [0, -1, 0, 1]
dy = [-1, 0, 1, 0]

def dfs(x, y):
    if dp[x][y]:
        return dp[x][y]
    else:
        dp[x][y] = 1
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n:
                if forest[nx][ny] > forest[x][y]:
                    dp[x][y] = max(dp[x][y], dfs(nx, ny) + 1)
        return dp[x][y]
        
def solve():
    global dp
    result = 0
    
    for i in range(n):
        for j in range(n):
            if not dp[i][j]:
                result = max(result, dfs(i, j))

    print(result)
            

if __name__ == '__main__':
    solve()